tests/test_qaoa.py

```diff
@@ -696,6 +707,121 @@ def test_wires_to_edges(self):
+                valid_bitstrings_indx.append(indx)
+            else:
+                invalid_bitstrings_indx.append(indx)
+
+        # Check that valid bitstrings map to a subset of the valid bitstrings
+        for indx in valid_bitstrings_indx:
+            column = h_matrix[:, indx]
+            destination_indxs = set(np.argwhere(column != 0).flatten())
+
+            assert destination_indxs.issubset(valid_bitstrings_indx)
+
+        # Check that invalid bitstrings map to a subset of the invalid bitstrings
+        for indx in invalid_bitstrings_indx:
+            column = h_matrix[:, indx]
+            destination_indxs = set(np.argwhere(column != 0).flatten())
+
+            assert destination_indxs.issubset(invalid_bitstrings_indx)
+
+        # Now consider a unitary generated by the Hamiltonian
+        h_matrix_e = expm(1j * h_matrix)
+
+        # We expect non-zero transitions among the set of valid bitstrings, and no transitions outside
+        for indx in valid_bitstrings_indx:
+            column = h_matrix_e[:, indx]
+            destination_indxs = np.argwhere(column != 0).flatten().tolist()
+            assert destination_indxs == valid_bitstrings_indx
+
+        # Check that invalid bitstrings transition within the set of invalid bitstrings
+        for indx in invalid_bitstrings_indx:
+            column = h_matrix_e[:, indx]
+            destination_indxs = set(np.argwhere(column != 0).flatten().tolist())
+            assert destination_indxs.issubset(invalid_bitstrings_indx)
+
+    def test_matrix(self):
+        """Test that the matrix function works as expected on a fixed example"""
+        g = nx.lollipop_graph(3, 1)
+        h = qml.qaoa.bit_flip_mixer(g, 0)
+
+        mat = matrix(h, 4)
+        mat_expected = np.array(
+            [
+                [0, 1, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0],
+                [1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0],
+                [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+                [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+                [1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+                [0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+                [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+                [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+                [1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0],
+                [0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0],
+                [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+                [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+                [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0],
+                [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0],
+                [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+                [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+            ]
+        )
+
+        assert np.allclose(mat.toarray(), mat_expected)
```

Comment:
I guess we can still keep this test even if we move `matrix` to the tests :thinking: